packages/dds/merge-tree/Obliterate.md

```diff
@@ -0,0 +1,276 @@
+
+Notice that the above examples always insert text at positions strictly inside the removed range.
+If the insert operation was instead before the "1" or after the "2", one can imagine different applications wanting different behavior:
+either the obliterated region should expand to include that text, or it should not.
+This topic will be covered in the [endpoint behavior](#endpoint-behavior) section,
+but for implementation strategy discussion one should assume that the implementation should support both options (and leave it up to applications
+to specify).
+
+## Implementation Strategy
+
+This section is focused on implementation of the "obliterate" semantics inside merge tree.
+This will constitute the bulk of the complexity of the feature.
+Since obliterate is generally a "different kind of remove," there may be a nice abstraction to introduce at the code level to generalize
+removal information. However, in favor of introducing niceties later this design document will assume fields are inlined and focus on
+the strategy for ensuring eventual consistency. If such an abstraction is introduced, ideally it would enable better "pay-to-play" of
+common code paths based on merge-tree feature usage.
+As an example, `BaseSegment.split()` needs to copy segment properties to the split segment.
+So new properties added to segment will unnecessarily copy undefined values.
+
+There are a few aspects of merge tree's bookkeeping and general feature set that require consideration when designing new op semantics:
+
+- Any changes to direct fields of tree nodes themselves (either new data or changes to bookkeeping of existing data)
+- How the feature interacts with an increasing collab window and zamboni
+- Impact on the partial lengths scheme
+- Bookkeeping and handling of overlapping removals (note some may be obliterates and some may not be)
+- Reconnection
+- Snapshotting impact
+
+We'll first present an overview of a potential scheme for implementing the obliterate op, then comment on these aspects.
+
+### Remote perspective
+
+First, consider the behavior a remote client must have when processing an obliterate op.
+For concreteness and ease of explanation, say this op is `{ seq: 50, refSeq: 40, clientId: 2, op: <obliterate the range [10, 15) }`.
+The processing client must first mark all segments between the segment `getContainingSegment({ pos: 10, refSeq: 40, clientId: 2 })` and
+`getContainingSegment({ pos: 15, refSeq: 40, clientId: 2 })` from the perspective `{ seq: 50, clientId: localClientId }` removed.
+The difference between this operation and a normal removal is its inclusion of segments inserted between seq 40 and seq 50.
+The current API on merge tree used for `markRangeRemoved` (which is `mapRange`) doesn't support iterating in this fashion,
+but could easily be extended to do so.
+One way to do that would be to decouple the `refSeq` and length calculations used for locating the positions and the `refSeq` used for
```

Comment:
there may be pre-existing removed segments in the tree in the obliterate range, some may be zamboni eligible, others not. how you handle those will be important for the insert design

Reply:
i think you are hinting at that here, but i think we should be explicit about their handling

Reply:
> how you handle those will be important for the insert design

which insert do you mean here? potential concurrent nearby ones?

pre-existing removed segments (i.e. removed before the refSeq of the obliterate op) should probably just be left alone. Ones that are concurrent to the obliterate op should have both fields (removedSeq, obliteratedSeq) set so that e.g. subsequent re-insertion that's still concurrent to the obliterate op correctly includes the new segment in the obliterate range. I hint at it in the next paragraph, but it's still probably a bit less explicit than it could be.

Reply:
when we do the excursion in an obliterated range that contains old deleted segments from before the obliterate how we do prevent those from causing the excursion to exit?

Reply:
Hm, I was planning on walking past them. This isn't implemented correctly in my pseudocode (there should be an early exit--we don't want to take the fact that it's been inserted much earlier as proof there hasn't been an adjacent obliterate since then), but it might also be correct to use the removedSeq as a lower bound for a potential obliterate/move. I need to think about that a bit more though.

Reply:
I changed my mind :). I think we only want to mark non-removed segments at sequencing time as obliterated. I adjusted the logic for the excursions to account for this. Maybe my note about using *all* of the information from removed segments is worth doing. Dunno. Basically, each segment that we visit that isn't moved tells us there was no move between `seg.seq` and `seg.removedSeq ?? currentSeq`. If we try to union all of these together via process of elimination, we'd want a structure unioning together disjoint intervals in the collab window. But probably any case where this gets really complex/inefficient is very degenerate in terms of having a large collab window.